File: LM 2016/MedSecret.java
import java.util.*;

public class MedSecret {
    public static void main(String [] args) {
        Scanner scan = new Scanner(System.in);
        int cases = Integer.parseInt(scan.nextLine());
        while (cases-- > 0) {
            // Read message.
            int m = Integer.parseInt(scan.nextLine());
            String [] lines = new String[m];
            for (int i = 0; i < m; i++) lines[i] = scan.nextLine();

            // Starting row, column.
            String [] yx = scan.nextLine().split(",");
            int row = Integer.parseInt(yx[0]);
            int col = Integer.parseInt(yx[1]);

            // Read cover.
            int co = Integer.parseInt(scan.nextLine());
            String [] cov = new String[co];
            for (int i = 0; i < co; i++) cov[i] = scan.nextLine();

            // Find secret.
            String secret = "";
            for (int r = 0; r < cov.length; r++) {
                // While there are still Os in the line.
                while (cov[r].indexOf('O') != -1) {
                    // Uncover correct letter and remove O from the line.
                    secret += lines[row + r].charAt(col + cov[r].indexOf('O'));
                    cov[r] = cov[r].replaceFirst("O", "-");
                }
            }
            System.out.println(secret);
        }
        scan.close();
    }
}

File: LM 2016/MedSecret.cpp
using namespace std;
#include <bits/stdc++.h>

// Read one int on line input.
int intline() {
    string res;
    getline(cin, res);
    return stoi(res);
}

// Return line input.
string strline() {
    string res;
    getline(cin, res);
    return res;
}

// Return split string.
vector<string> split(string line, string delim) {
    vector<string> res;
    while (line.find(delim) != string::npos) {
        res.push_back(line.substr(0, line.find(delim)));
        line = line.substr(line.find(delim) + delim.length());
    }
    res.push_back(line);
    return res;
}

// Return int version of string vector.
vector<int> strToInt(vector<string> list) {
    vector<int> res;
    for (auto &x : list) res.push_back(stoi(x));
    return res;
}

int main()
{
    // Test cases.
    int cases = intline();
    while (cases-- > 0) {
        // Read message.
        int m = intline();
        vector<string> lines(m);
        for (auto &x : lines) x = strline();
        
        // Starting row, column.
        vector<int> yx = strToInt(split(strline(), ","));
        int y = yx[0], x = yx[1];
        
        // Read cover.
        int c = intline();
        vector<string> cov(c);
        for (auto &x : cov) x = strline();
        
        // Find secret.
        string sec = "";
        for (int r = 0; r < cov.size(); r++) {
            while (cov[r].find('O') != string::npos) {
                int pos = cov[r].find('O');
                sec += lines[y + r][x + pos];
                cov[r][pos] = '-';
            }
        }
        
        // Result.
        cout << sec << endl;
    }

    return 0;
}
